near-plugins-derive/src/access_controllable.rs

```diff
@@ -0,0 +1,287 @@
+                    };
+                    event.emit();
+                }
+
+                is_new_grantee
+            }
+
+            fn has_role(&self, role: #role_type, account_id: &::near_sdk::AccountId) -> bool {
+                match self.permissions.get(account_id) {
+                    Some(permissions) => {
+                        let flag = <#bitflags_type>::from_bits(role.acl_permission())
+                            .expect(#ERR_PARSE_BITFLAG);
+                        permissions.contains(flag)
+                    }
+                    None => false,
+                }
+            }
+
+            fn has_any_role(
+                &self, roles: Vec<#role_type>,
+                account_id: &::near_sdk::AccountId
+            ) -> bool {
+                let permissions: Vec<#bitflags_type> = roles
+                    .iter()
+                    .map(|role| {
+                        <#bitflags_type>::from_bits(role.acl_permission())
+                            .expect(#ERR_PARSE_BITFLAG)
+                    })
+                    .collect();
+                let target = permissions.iter().fold(
+                    <#bitflags_type>::empty(),
+                    |acc, &x| acc | x,
+                );
+                self.has_any_permission(target, account_id)
+            }
+
+            fn has_any_permission(&self, target: #bitflags_type, account_id: &::near_sdk::AccountId) -> bool {
+                let permissions = self.get_or_init_permissions(account_id);
+                target.intersects(permissions)
+            }
+
+            /// Adds `account_id` to the set of `permission` bearers.
+            fn add_bearer(&mut self, permission: #bitflags_type, account_id: &::near_sdk::AccountId) {
+                let mut set = match self.bearers.get(&permission) {
+                    Some(set) => set,
+                    None => Self::new_bearers_set(permission),
+                };
+                if let true = set.insert(account_id) {
+                    self.bearers.insert(&permission, &set);
+                }
+            }
+        }
+
+        // Note that `#[near-bindgen]` exposes non-public functions in trait
+        // implementations. This is [documented] behavior. Therefore some
+        // functions are made `#[private]` despite _not_ being public.
+        //
+        // [documented]: https://docs.near.org/sdk/rust/contract-interface/public-methods#exposing-trait-implementations
+        #[near_bindgen]
+        impl AccessControllable for #ident {
+            fn acl_storage_prefix() -> &'static [u8] {
+                (#storage_prefix).as_bytes()
+            }
+
+            #[private]
+            fn acl_grant_role_unchecked(&mut self, role: String, account_id: ::near_sdk::AccountId) -> bool {
+                let role = <#role_type>::try_from(role.as_str()).expect(#ERR_PARSE_ROLE);
+                self.#acl_field.grant_role_unchecked(role, &account_id)
```

Comment:
Can you remind me what was the reason to have acl as an injected field versus having a function that builds this object dynamically? For example:

```suggestion
                self.__acl_object().grant_role_unchecked(role, &account_id)
```

Reply:
The part I don't like about injecting the code without letting developers know is that they might struggle to upgrade their contracts and serialize/deserialize the state later.

Reply:
If the field is not injected we (or the developers) have to manually (de)serialize and read/write the object from/to storage. Probably using [`storage_read`](https://docs.rs/near-sdk/latest/near_sdk/env/fn.storage_read.html) and [`storage_write`](https://docs.rs/near-sdk/latest/near_sdk/env/fn.storage_write.html). We discussed offline to inject the field, to avoid the steps mentioned above.

---

I have not yet checked if it's feasible, but another approach might be:

- Ask the developer to have a field `__acl: ()`. The type this field should ultimately have is defined only inside the attribute macro. So I think developers can't write `__acl: AclType` in their contract.
- Inside the macro, we change the type of this field (instead of injecting a new field).

Probably it wouldn't be a huge win, but makes it more explicit.

Reply:
[`near-floating-state`](https://github.com/mfornet/near-floating-state) fits nicely here! Let's stay with injection for now to finalize this PR soon? Then once `near-floating-state` is on crates.io we can introduce it for Acl in a follow-up PR.